source/opt/inline_pass.cpp

```diff
@@ -402,23 +468,117 @@ bool InlinePass::Inline(ir::Function* func) {
   return modified;
 }
 
-bool InlinePass::IsInlinableFunction(const ir::Function* func) {
-  // We can only inline a function if it has blocks.
-  if (func->cbegin() == func->cend())
+bool InlinePass::hasMultipleReturns(ir::Function* func) {
+  bool seenReturn = false;
+  bool multipleReturns = false;
+  for (auto& blk : *func) {
+    auto lii = blk.cend();
+    lii--;
+    if (lii->opcode() == SpvOpReturn || lii->opcode() == SpvOpReturnValue) {
+      if (seenReturn) {
+        multipleReturns = true;
+        break;
+      }
+      seenReturn = true;
+    }
+  }
+  return multipleReturns;
+}
+
+void InlinePass::ComputeStructuredSuccessors(ir::Function* func) {
+  // If header, make merge block first successor.
+  for (auto& blk : *func) {
+    auto mii = blk.end();
+    mii--;
+    uint32_t mbid = 0;
+    if (mii != blk.begin()) {
+      mii--;
+      if (mii->opcode() == SpvOpLoopMerge)
+        mbid = mii->GetSingleWordOperand(kSpvLoopMergeMergeBlockId);
+      else if (mii->opcode() == SpvOpSelectionMerge)
+        mbid = mii->GetSingleWordOperand(kSpvSelectionMergeMergeBlockId);
+    }
```

Comment:
Optional: This code sequence could have been clearer if it were factored out into a function:  
```
// Returns the id of the merge block declared by a merge instruction in this block, if any.  If none, returns zero.
uint32_t MergeBlockIdIfAny(const ir::BasicBlock&);
```